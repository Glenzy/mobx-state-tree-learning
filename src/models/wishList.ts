import { types, Instance, getParent, destroy } from 'mobx-state-tree';
import { entries } from 'mobx';

const data = {
  name: 'The Chronicals Of Narnia - CS Lewis Box Set',
  price: 19.99,
  image:
    'https://www.booktopia.com.au/http_coversbooktopiacomau/big/9780007528097/0000/the-chronicles-of-narnia-the-chronicles-of-narnia-boxed-set.jpg'
};

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: types.maybe(types.string)
  })
  .actions(self => ({
    changeName(newName: string) {
      self.name = newName;
    },
    changePrice(newPrice: number) {
      self.price = newPrice;
    },
    changeImage(newImage: string) {
      self.image = newImage;
    },
    removeItem() {
      (getParent(self, 2) as IWishList).removeItem(self as IWishListItem);
    }
  }));



export type IWishListItem = Instance<typeof WishListItem>;

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => {
    function addItem(item: IWishListItem) {
      self.items.push(item);
      addItemToLocalStorage(item);
    }
    function addItemToLocalStorage(item: IWishListItem) {
      let localStorageItem;
      //let key: any
      for (let key in item) {
        if (item.hasOwnProperty(key)) {
          localStorageItem = item[key];
          console.log(typeof (item[key]));
        }
      }
    }
    function removeItem(item: IWishListItem) {
      destroy(item);
    }
    return {
      addItem,
      addItemToLocalStorage,
      removeItem
    }
  })
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum: any, entry: any) => sum + entry.price, 0);
    }
  }));
export type IWishList = Instance<typeof WishList>;
