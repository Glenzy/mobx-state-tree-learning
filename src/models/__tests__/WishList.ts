import { getSnapshot, onSnapshot } from 'mobx-state-tree';
import { reaction } from 'mobx';

import { WishListItem, WishList } from '../wishList';

function crateWishList() {
    const list = WishList.create({
        items: [
            {
                name: "The Chronicals Of Narnia - CS Lewis Box Set",
                price: 19.99,
            }
        ]
    });

    return list
}

it('should create an instance of a model', () => {
    const item = WishListItem.create({
        "name": "The Chronicals Of Narnia - CS Lewis Box Set",
        "price": 19.99,
    });
    expect(item.price).toBe(19.99);
    expect(item.image).toBe('');
});

it('should create a WishList', () => {
    const list = crateWishList();
    expect(list.items.length).toBe(1);
    expect(list.items[0].price).toBe(19.99);
});


it('should change the item name', () => {
    const list = crateWishList();
    list.items[0].changeName("Moo");
    expect(getSnapshot(list)).toMatchSnapshot()
})


it('should change the item price', () => {
    const list = crateWishList();
    list.items[0].changePrice(21.00);
    expect(getSnapshot(list)).toMatchSnapshot()
})


it('should add a new item and update price', () => {
    const list = crateWishList();
    const states: [] = [];
    onSnapshot(list, (snapshot: never) => {
        states.push(snapshot);
    })

    list.addItem({
        name: "The Game Of Thrones Ninja Edition",
        price: 21.99
    })
    list.items[1].changePrice(22.00);

    expect(getSnapshot(list)).toMatchSnapshot()
    expect(states).toMatchSnapshot()
})

it('can calculate the total price of the list', () => {
    const list = crateWishList();
    list.addItem({
        name: "The Game Of Thrones Ninja Edition",
        price: 21
    })
    list.addItem({
        name: "Ninjas Will Take Your Lunch",
        price: 100
    })

    expect(list.totalPrice).toBe(140.99);
})

it('will not mutate other values', () => {
    const list = crateWishList();
    list.addItem({
        name: "The Game Of Thrones Ninja Edition",
        price: 21
    })
    list.addItem({
        name: "Ninjas Will Take Your Lunch",
        price: 100
    })
    let changed: number = 0;
    reaction(() => list.totalPrice, () => changed++);
    list.items[0].changeImage('http://some-image-url.com/images/image');
    list.items[1].changeName('Game Of Thrones - Winter is Coming');
    expect(changed).toBe(0);
    list.items[1].changePrice(500);
    expect(changed).toBe(1);
})