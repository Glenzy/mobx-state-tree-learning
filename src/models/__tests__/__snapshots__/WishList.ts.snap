// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a new item and update price 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "The Chronicals Of Narnia - CS Lewis Box Set",
      "price": 19.99,
    },
    Object {
      "image": "",
      "name": "The Game Of Thrones Ninja Edition",
      "price": 22,
    },
  ],
}
`;

exports[`should add a new item and update price 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "The Chronicals Of Narnia - CS Lewis Box Set",
        "price": 19.99,
      },
      Object {
        "image": "",
        "name": "The Game Of Thrones Ninja Edition",
        "price": 21.99,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "The Chronicals Of Narnia - CS Lewis Box Set",
        "price": 19.99,
      },
      Object {
        "image": "",
        "name": "The Game Of Thrones Ninja Edition",
        "price": 22,
      },
    ],
  },
]
`;

exports[`should change the item name 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Moo",
      "price": 19.99,
    },
  ],
}
`;

exports[`should change the item price 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "The Chronicals Of Narnia - CS Lewis Box Set",
      "price": 21,
    },
  ],
}
`;
